### PR TITLE
Bugfix/ml 402 fix tensor reshape

### PR DESCRIPTION
**Pull request by baykaner**
- fixes reshape op to work as expected

### CODE REVIEW

File: libs/math/tests/unit/tensor/tensor_operations.cpp [hunk 1]
```diff
@@ -21,21 +21,85 @@
...
+  ReshapeTest<TypeParam>({2, 1, 3, 1}, {6, 1, 1});
+  ReshapeTest<TypeParam>({2, 1, 1, 3}, {2, 1, 3});
+  ReshapeTest<TypeParam>({2, 3, 4, 5}, {12, 2, 5});
+
+  ReshapeTest<TypeParam>({6, 1, 1}, {2, 3, 1, 1});
+  ReshapeTest<TypeParam>({2, 1, 3}, {2, 1, 3, 1});
+  ReshapeTest<TypeParam>({1, 1, 6}, {2, 1, 1, 3});
+  ReshapeTest<TypeParam>({3, 4, 10}, {2, 3, 4, 5});
+}
+
```
Comment (vhavrylov-ncube) [inline libs/math/tests/unit/tensor/tensor_operations.cpp:100]: What about negative cases, when reshape should fail/throw ?

File: libs/ml/include/ml/ops/reshape.hpp [hunk 2]
```diff
@@ -40,12 +40,15 @@ class Reshape : public fetch::ml::ops::Ops<T>
 
   explicit Reshape(std::vector<SizeType> new_shape)
     : new_shape_(std::move(new_shape))
-  {}
+  {
+    new_size_ = fetch::math::Product(new_shape_);
```
Comment (vhavrylov-ncube) [inline libs/ml/include/ml/ops/reshape.hpp]: Wouldn't it be better to place this in initializer list?

File: libs/ml/include/ml/ops/reshape.hpp [hunk 3]
```diff
@@ -40,12 +40,15 @@ class Reshape : public fetch::ml::ops::Ops<T>
...
+  }
 
   explicit Reshape(SPType const &sp)
     : Ops<T>(sp)
   {
     new_shape_ = sp.new_shape;
+    new_size_  = sp.new_size;
```
Comment (vhavrylov-ncube) [inline libs/ml/include/ml/ops/reshape.hpp]: initialiser list again

File: libs/ml/tests/unit/ops/reshape.cpp [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+      new_op.Backward({std::make_shared<const TensorType>(data)}, error);
+
+  // test correct values
+  EXPECT_TRUE(error_signal.at(0).AllClose(
+      new_error_signal.at(0), fetch::math::function_tolerance<typename TypeParam::Type>(),
+      fetch::math::function_tolerance<typename TypeParam::Type>()));
+  fetch::math::state_clear<DataType>();
+}
+
+TYPED_TEST(ReshapeTest, Reshape_graph_serialization_test)
```
Comment (vhavrylov-ncube) [inline libs/ml/tests/unit/ops/reshape.cpp]: serialisation

File: libs/ml/tests/unit/ops/reshape.cpp [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+
+  data.FillUniformRandom();
+  auto gt_error_it = gt_error.begin();
+  auto data_it     = data.begin();
+  for (std::size_t j = 0; j < gt_error.size(); ++j)
+  {
+    *gt_error_it = *data_it;
+    ++gt_error_it;
+    ++data_it;
+  }
```
Comment (evsmithx) [inline libs/ml/tests/unit/ops/reshape.cpp]: I think you can use gt_error = data.Copy() here.

**Review by evsmithx — COMMENTED** [review]
There doesn't seem to be a test here for the order that the elements are put into the new tensor, i.e. whether `[1, 2, 3, 4].reshape(2,2)` gives [[1, 2], [3, 4]] or [[1, 3], [2, 4]].

File: libs/math/tests/unit/tensor/tensor_operations.cpp (hunk 1 above)
Comment (evsmithx) [reply libs/math/tests/unit/tensor/tensor_operations.cpp:100]: The parameters are vectors of SizeTypes, so presumably it would give a compilation error.

**Review by vhavrylov-ncube — DISMISSED** [review]
Looks good to me

**Review by evsmithx — CHANGES_REQUESTED** [review]
Batch dimension of tensor in reshape op needs to stay the same. Tests for this and also for ComputeOutputShape needed, as discussed on slack.

**Review by evsmithx — APPROVED** [review]

**Pull request merged**